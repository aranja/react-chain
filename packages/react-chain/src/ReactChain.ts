import createBase from './ReactChainBase'
import { ReactElement } from 'react'
import Session from './Session'
import { getChainForTarget, renderRecursively, render } from './SessionUtils'

export type RenderTarget =
  'browser' |
  'server'

export type WrapElement =
  (renderChildren: () => Promise<null | ReactElement<any>>, context: any) =>
    ReactElement<any> | Promise<ReactElement<any>>

export type WrapRender =
  (render: Function) =>
    void

export type Middleware =
  (session: Session) =>
    (void | WrapElement)

export class ReactChain {
  protected middlewareChain: Array<Middleware> = []

  chain(middleware?: Middleware) {
    if (typeof middleware !== 'function') {
      throw new Error('A react-chain middleware should be a function')
    }

    this.middlewareChain.push(middleware)

    return this
  }

  async getElement(session = new Session()): Promise<any> {
    if (session.__firstRender) {
      session.__elementChain = [createBase]
      this.middlewareChain.forEach(middleware => {
        const createElement = middleware(session)
        if (createElement) {
          session.__elementChain.push(createElement)
        }
      })
      session.__firstRender = false
    }

    let index = 0
    const next = (): Promise<any> => {
      const createElement = session.__elementChain[index++]
      const renderChildren = session.__elementChain[index]
        ? next
        : () => Promise.resolve(null)
      return Promise.resolve(createElement(renderChildren, session))
    }

    return await next()
  }

  renderBrowser(session: Session, onRender: () => void) {
    render(session, 'browser')(onRender)
  }

  renderServer(session: Session, onRender: Function) {
    let body = ''
    render(session, 'server')(() => {
      body = onRender()
    })
    return body
  }
}

export default function createReactChain() {
  return new ReactChain()
}
