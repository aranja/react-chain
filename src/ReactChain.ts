import createSession from './Session'
import { isValidElement, ReactElement } from 'react'
import { validateElementCreator, unfoldRender, renderElementChain } from './utils'
import { Middleware, Session, CreateElement } from './types'
export { Middleware, Session, CreateElement }

export class ReactChain {
  middlewareChain: Array<Middleware> = []

  use(middleware?: ReactElement<any> | Middleware) {
    if (middleware && isValidElement(middleware)) {
      middleware = ((element) => () => () => element)(middleware)
    }
    if (typeof middleware !== 'function') {
      throw new Error('A react-chain middleware should be a function')
    }

    this.middlewareChain.push(middleware)

    return this
  }

  // Deprecated
  chain(middleware?: ReactElement<any> | Middleware) {
    return this.use(middleware)
  }

  createSession = createSession

  async getElement(session?: Session): Promise<any> {
    if (session == null) {
      throw new Error('Missing session object.')
    }

    if (session.__firstRender) {
      this.middlewareChain.forEach(middleware => {
        let createElement: void | CreateElement
        if ((createElement = validateElementCreator(middleware(session)))) {
          session.__elementChain.push(createElement)
        }
      })

      session.__firstRender = false
    }

    return await renderElementChain(session.__elementChain)
  }

  async renderBrowser(session: Session, onRender: (element: ReactElement<any>) => any) {
    const element = await this.getElement(session)
    return unfoldRender(session, 'browser', () => onRender(element))
  }

  async renderServer(session: Session, onRender: (element: ReactElement<any>) => any) {
    const element = await this.getElement(session)
    return unfoldRender(session, 'server', () => onRender(element))
  }
}

export default function createReactChain() {
  return new ReactChain()
}
