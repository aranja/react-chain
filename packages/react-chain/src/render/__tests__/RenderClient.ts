import renderClient from '../RenderClient'
import createReactChain from '../../ReactChain'
import { ReactChain } from '../../ReactChain'
import { createElement } from 'react'

describe('renderClient()', () => {
  let appRoot: Element
  let app: ReactChain

  beforeEach(() => {
    app = createReactChain()
    appRoot = document.createElement('div')
    document.body.appendChild(appRoot)
  })

  it('should be callable', () => {
    expect(typeof renderClient).toBe('function')
  })

  it('should return a promise', () => {
    const client = renderClient(app, appRoot)
    expect(typeof client.then).toBe('function')
    expect(typeof client.catch).toBe('function')
  })

  it('should await getElement before rendering', async () => {
    let runOrder = 0
    app.getElement = jest.fn(() => { runOrder = 2 })
    app.renderClient = jest.fn(() => { runOrder *= 4 })
    await renderClient(app, appRoot)
    expect(app.getElement).toHaveBeenCalled()
    expect(app.renderClient).toHaveBeenCalled()
    expect(runOrder).toBe(8)
  })

  it('should add a refresh function on the context', async () => {
    let context = { refresh: null }
    app.getElement = jest.fn((internalContext) => { context = internalContext })
    app.renderClient = jest.fn()
    await renderClient(app, appRoot)
    expect(typeof context.refresh).toBe('function')
  })

  it('should render a React component to the appRoot', async () => {
    app.chain(() => createElement('div', {}, 'React Element'))
    await renderClient(app, appRoot)
    expect(appRoot.innerHTML).toMatchSnapshot()
  })

  it('should rerender when refresh is called', (done) => {
    let renderCount = 0

    function onComplete() {
      expect(renderCount).toBe(2)
      done()
    }

    app.chain(() => createElement('div', {}, `Render count ${++renderCount}`))

    app.chain({
      wrapClientRender(render, context) {
        if (renderCount === 1 && context.refresh) {
          context.refresh(onComplete)
        }
        render()
      }
    })

    renderClient(app, appRoot)
  })
})