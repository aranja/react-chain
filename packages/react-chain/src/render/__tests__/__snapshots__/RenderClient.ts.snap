// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderClient() should render a React component to the appRoot 1`] = `"<div data-reactroot=\\"\\">React Element</div>"`;
